### ISSUE TITLE
GDAL chokes opening attached JPG, yet works OK in other applications

### ISSUE THREAD

**Issue by nyalldawson** [OSGeo/gdal#1545]
## Expected behavior and actual behavior.

The attached JPG file triggers errors when attempting to convert using gdal_translate:

```
Input file size is 5728, 4902
0ERROR 6: Reading this image would require libjpeg to allocate at least 252707904 bytes. This is disabled since above the 104857600 threshold. You may override this restriction by defining the GDAL_ALLOW_LARGE_LIBJPEG_MEM_ALLOC environment variable, or recompile GDAL by defining the GDAL_LIBJPEG_LARGEST_MEM_ALLOC macro to a value greater than 104857600
ERROR 1: 32-o_2016.jpg, band 1: IReadBlock failed at X offset 0, Y offset 0: Reading this image would require libjpeg to allocate at least 252707904 bytes. This is disabled since above the 104857600 threshold. You may override this restriction by defining the GDAL_ALLOW_LARGE_LIBJPEG_MEM_ALLOC environment variable, or recompile GDAL by defining the GDAL_LIBJPEG_LARGEST_MEM_ALLOC macro to a value greater than 104857600
```

Yet the same jpg opens without issues in other applications (e.g. GIMP).

## Steps to reproduce the problem.

Unzip attached file, run `gdal_translate 32-o_2016.jpg -of "GTIFF" test.TIF`

## Operating system

Fedora 29

## GDAL version and provenance

master (6ad7dccaf6aa2dba826d160df552a982bd9a406a), reproduced on 2.3.2 also. 

[32-o_2016.zip](https://github.com/OSGeo/gdal/files/3196314/32-o_2016.zip)

**Comment by rouault** [comment]
This image uses JPEG features that might require up to (width * height * 3 * 3) bytes to be allocated by libjpeg to be decoded, which could be used by an attacker to DoS the application. See LJT­01­004 of http://www.libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf
As the error message suggest defining GDAL_ALLOW_LARGE_LIBJPEG_MEM_ALLOC=YES as env variable / config option will remove that protection.

**Comment by nyalldawson** [comment]
Thanks @rouault  -- I can confirm that GDAL_ALLOW_LARGE_LIBJPEG_MEM_ALLOC works ok for this one (and it's something which needs to be caught in QGIS, instead of returning garbage data which is the current result).

I was thrown a bit since googling for returns almost nothing, so wasn't sure if it was an outdated message getting thrown as a result of some other error.

**Comment by Saijin-Naib** [comment]
I'll ask in here then:
Why is the behavior different under Linux? That seems like a bug.

**Comment by enzedrailmaps** [comment]
Just seeing this for the first time, which I think is from upgrading 3.22.4 to 3.22.5.

2022-04-18T01:53:51     WARNING    D1961_94CFT_92XGM_94CFW_92XGN_f6e10828_d0fa_42e6_a318_901bef99006f :: /mnt/sharerw/mainpc/maps/References/Volume 11-12 - MSL/Station - Dunsandel/D1961:94CFT-92XGM+94CFW-92XGN.jpg, band 1: IReadBlock failed at X offset 0, Y offset 600: Reading this image would require libjpeg to allocate at least 1866240000 bytes. This is disabled since above the 104857600 threshold. You may override this restriction by defining the GDAL_ALLOW_LARGE_LIBJPEG_MEM_ALLOC environment variable, or recompile GDAL by defining the GDAL_LIBJPEG_LARGEST_MEM_ALLOC macro to a value greater than 104857600

The file is 26 MB so this makes no sense as the numbers translate into 1 GB.

**Comment by rouault** [comment]
> The file is 26 MB so this makes no sense as the numbers translate into 1 GB.

Not necessarily... Some pathological JPEG formulations might require a lot of memory to decompress. Do you have a link to that file ?

**Comment by Saijin-Naib** [comment]
> > The file is 26 MB so this makes no sense as the numbers translate into 1 GB.
> 
> Not necessarily... Some pathological JPEG formulations might require a lot of memory to decompress. Do you have a link to that file ?

I have one attached to my older/closed issue with the same behavior: 
https://github.com/qgis/QGIS/issues/34558

**Comment by rouault** [comment]
> I have one attached to my older/closed issue with the same behavior:
> [qgis/QGIS#34558](https://github.com/qgis/QGIS/issues/34558)

that image only requires 37 MB, which is below the 500 MB threshold used by the JPEG driver

> " or recompile GDAL by defining the GDAL_LIBJPEG_LARGEST_MEM_ALLOC macro to a value greater than 104857600"

This message is intriguing. It doesn't come from the JPEG driver, but the JPEG plugin of the MRF driver. It is weird that the MRF driver kicks in here.

**Comment by agiudiceandrea** [comment]
It seems to me there is also an oversight in the warning message:
"Reading this image would require libjpeg to <ins>allocate at least **1866240000** bytes</ins>. This is disabled since above the **104857600** threshold. You may override this restriction by ... recompile GDAL by defining the GDAL_LIBJPEG_LARGEST_MEM_ALLOC macro to a value <ins>greater than **104857600**</ins>".

Actually the GDAL_LIBJPEG_LARGEST_MEM_ALLOC macro should be defined to a value **equal or greater then** the required memory, i.e **1866240000** bytes (not 104857600 bytes) ", to avoid the error.

**Comment by rouault** [comment]
> It seems to me there is also an oversight in the warning message:

indeed. CC @lucianpls . But the MRF driver shouldn't really kick in on a simple JPEG image

**Comment by lucianpls** [comment]
It was never meant to.

**Comment by lucianpls** [comment]
It doesn't use the MRF driver.
It could only happen if an MRF dataset would be purposefully built around the JPEG.

**Comment by enzedrailmaps** [comment]
The issue I am seeing is with using a Progressive JPEG exported from Gimp. No progressive setting, no problems in Qgis.
You should be easily able to replicate this yourself.

**Comment by rouault** [comment]
I could reproduce something similar with a 20000 x 15000 pixels JPEG image.
The message I get is
"ERROR 6: Reading this image would require libjpeg to allocate at least 1801048576 bytes. This is disabled since above the 524288000 threshold. You may override this restriction by defining the GDAL_ALLOW_LARGE_LIBJPEG_MEM_ALLOC environment variable, or setting the JPEGMEM environment variable to a value greater or equal to '1802M'"
(the message you got makes me thing you might use GDAL < 3.0)
This is correct. Decoding such image requires 1.8 GB of temporary memory. Following the hints allows to decode it. https://github.com/qgis/QGIS/issues/35982 has been filed about possibly setting GDAL_ALLOW_LARGE_LIBJPEG_MEM_ALLOC in QGIS.

**Comment by enzedrailmaps** [comment]
Sample image:
https://drive.google.com/drive/folders/1mU7EyBQCCxrRhxHd6yLUoei9Nf_d7hLd?usp=sharing

**Comment by enzedrailmaps** [comment]
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>


QGIS version | 3.22.5-Białowieża | QGIS code revision | c27231782f
-- | -- | -- | --
Qt version | 5.12.8
Python version | 3.8.10
GDAL/OGR version | 3.0.4
PROJ version | 6.3.1
EPSG Registry database version | v9.8.6 (2020-01-22)
Compiled against GEOS | 3.8.0-CAPI-1.13.1 | Running against GEOS | 3.8.0-CAPI-1.13.1
SQLite version | 3.31.1
PDAL version | 2.0.1
PostgreSQL client version | 12.9 (Ubuntu 12.9-0ubuntu0.20.04.1)
SpatiaLite version | 4.3.0a
QWT version | 6.1.4
QScintilla2 version | 2.11.2
OS version | Ubuntu 20.04.4 LTS
  |   |   |  
Active Python plugins
grassprovider | 2.12.99
db_manager | 0.1.20
sagaprovider | 2.12.99
processing | 2.12.99
MetaSearch | 0.3.5

</body></html>QGIS version
3.22.5-Białowieża
QGIS code revision
[c27231782f](https://github.com/qgis/QGIS/commit/c27231782f)
Qt version
5.12.8
Python version
3.8.10
GDAL/OGR version
3.0.4
PROJ version
6.3.1
EPSG Registry database version
v9.8.6 (2020-01-22)
Compiled against GEOS
3.8.0-CAPI-1.13.1
Running against GEOS
3.8.0-CAPI-1.13.1 
SQLite version
3.31.1
PDAL version
2.0.1
PostgreSQL client version
12.9 (Ubuntu 12.9-0ubuntu0.20.04.1)
SpatiaLite version
4.3.0a
QWT version
6.1.4
QScintilla2 version
2.11.2
OS version
Ubuntu 20.04.4 LTS




Active Python plugins
grassprovider
2.12.99
db_manager
0.1.20
sagaprovider
2.12.99
processing
2.12.99
MetaSearch
0.3.5

**Comment by enzedrailmaps** [comment]
I think the question to be addressed is how it is that a greater memory allocation is required by GDAL for a progressive JPEG. I never saw any issues like this before with previous versions of either Qgis/GDAL that didn't create the problems for a progressive JPEG. I am not sure why this issue has developed. But I can see that there could end up being a problem with a large memory allocation requirement.

**Comment by rouault** [comment]
ok GDAL 3.0.4 has indeed the old error message in the JPEG driver (I was looking at the sources of the latest state of the 3.0 branch that has the new message released in 3.1.0 and later), which is now only existing in the MRF driver. "mystery" solved for that part

> I think the question to be addressed is how it is that a greater memory allocation is required by GDAL for a progressive JPEG.

Whole decompression of progressive JPEG is required by libjpeg itself (and basically a constraint of the format itself), and it is completely expected. The current behavior of trying to anticipate the huge memory consumption is what is recommended by the authors of libjpeg to avoid denial of service attacks.

>  I never saw any issues like this before with previous versions of either Qgis/GDAL

The detection of the huge memory allocation dates from GDAL 2.3. Older version would consume a lot of memory too but without detecting it.
